frontend/src/metabase/actions/containers/ActionCreator/ActionCreator.tsx

```diff
@@ -0,0 +1,220 @@
+import type Question from "metabase-lib/Question";
+
+import CreateActionForm from "../CreateActionForm";
+import ActionCreatorHeader from "./ActionCreatorHeader";
+import QueryActionEditor from "./QueryActionEditor";
+import FormCreator from "./FormCreator";
+import {
+  DataReferenceTriggerButton,
+  DataReferenceInline,
+} from "./InlineDataReference";
+
+import {
+  ActionCreatorBodyContainer,
+  EditorContainer,
+  ModalRoot,
+  ModalActions,
+  ModalLeft,
+} from "./ActionCreator.styled";
+
+import { newQuestion } from "./utils";
+
+const mapStateToProps = (
+  state: State,
+  { action }: { action: WritebackQueryAction },
+) => ({
+  metadata: getMetadata(state),
+  question: action ? createQuestionFromAction(state, action) : undefined,
+  actionId: action ? action.id : undefined,
+});
+
+const mapDispatchToProps = {
+  push,
+  update: Actions.actions.update,
+};
+
+interface ActionCreatorProps {
+  metadata: Metadata;
+  question?: Question;
+  actionId?: number;
+  modelId?: number;
+  databaseId?: number;
+  push: (url: string) => void;
+  update: (action: ActionParams) => void;
+  onClose?: () => void;
+}
+
+function ActionCreatorComponent({
+  metadata,
+  question: passedQuestion,
+  actionId,
+  modelId,
+  databaseId,
+  update,
+  onClose,
+}: ActionCreatorProps) {
+  const [question, setQuestion] = useState(
+    passedQuestion ?? newQuestion(metadata, databaseId),
+  );
+  const [formSettings, setFormSettings] = useState<
+    ActionFormSettings | undefined
+  >(undefined);
+  const [showSaveModal, setShowSaveModal] = useState(false);
+
+  const [isDataRefOpen, { toggle: toggleDataRef, turnOff: closeDataRef }] =
+    useToggle(false);
+
+  useEffect(() => {
+    setQuestion(passedQuestion ?? newQuestion(metadata, databaseId));
+
+    // we do not want to update this any time the props or metadata change, only if action id changes
+  }, [actionId]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const defaultModelId: number | undefined = useMemo(() => {
+    if (modelId) {
+      return modelId;
+    }
+    const params = new URLSearchParams(window.location.search);
+    const modelQueryParam = params.get("model-id");
+    return modelId ? Number(modelQueryParam) : undefined;
+  }, [modelId]);
+
+  if (!question || !metadata) {
+    return null;
+  }
+
+  const query = question.query() as NativeQuery;
+
+  const isNew = !actionId && !(question.card() as SavedCard).id;
+
+  const handleSaveClick = () => {
+    if (isNew) {
+      setShowSaveModal(true);
+    } else {
+      update({
+        id: (question.card() as SavedCard).id,
+        name: question?.displayName() ?? "",
+        description: question.description() ?? null,
+        model_id: defaultModelId as number,
+        formSettings: formSettings as ActionFormSettings,
+        question,
+      });
+      onClose?.();
+    }
+  };
+
+  const handleOnSave = (action: WritebackQueryAction) => {
+    setQuestion(question.setCard(action as unknown as SavedCard));
```

Comment:
nit: looks like something needs to be changed in the action type but I'm not sure

Reply:
there's probably a better way to handle this - this is a consequence of using the query editor which assumes we're editing a `Question`, and actions were originally a type of question, but have now been separated from them.

Reply:
Extract this to a utility function, hopefully it'd make it a bit more straightforward